File: src/test/mocks/fixtures.ts
import type {
  Account,
  Budget,
  Category,
  DashboardSummary,
  RecurringBill,
  SavingsGoal,
  Transaction,
} from "../../types/wallet";

export const accountFixture: Account = {
  id: "account-1",
  name: "Checking",
  accountType: "cash",
  currency: "MAD",
  initialBalanceMinor: 100_000,
  balanceMinor: 125_000,
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const incomeCategoryFixture: Category = {
  id: "category-income",
  name: "Salary",
  categoryType: "income",
  icon: null,
  color: null,
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const expenseCategoryFixture: Category = {
  id: "category-expense",
  name: "Groceries",
  categoryType: "expense",
  icon: null,
  color: null,
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const incomeTransactionFixture: Transaction = {
  id: "transaction-income",
  accountId: accountFixture.id,
  categoryId: incomeCategoryFixture.id,
  transactionType: "income",
  amountMinor: 200_000,
  description: "June salary",
  transactionDate: "2026-06-10",
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const expenseTransactionFixture: Transaction = {
  id: "transaction-expense",
  accountId: accountFixture.id,
  categoryId: expenseCategoryFixture.id,
  transactionType: "expense",
  amountMinor: 12_500,
  description: "Weekly groceries",
  transactionDate: "2026-06-09",
  createdAt: "2026-06-09T00:00:00Z",
  updatedAt: "2026-06-09T00:00:00Z",
};

export const budgetFixture: Budget = {
  id: "budget-1",
  name: "Food budget",
  categoryId: expenseCategoryFixture.id,
  categoryName: expenseCategoryFixture.name,
  amountMinor: 100_000,
  spentMinor: 25_000,
  remainingMinor: 75_000,
  progressPercentage: 25,
  isNearLimit: false,
  isExceeded: false,
  month: 6,
  year: 2026,
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const savingsGoalFixture: SavingsGoal = {
  id: "goal-1",
  name: "Emergency fund",
  targetAmountMinor: 500_000,
  currentAmountMinor: 125_000,
  remainingAmountMinor: 375_000,
  progressPercent: 25,
  deadlineDate: "2026-12-31",
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const recurringBillFixture: RecurringBill = {
  id: "bill-1",
  name: "Internet",
  accountId: accountFixture.id,
  accountName: accountFixture.name,
  categoryId: expenseCategoryFixture.id,
  categoryName: expenseCategoryFixture.name,
  amountMinor: 25_000,
  frequency: "monthly",
  nextDueDate: "2026-06-15",
  lastPaidDate: null,
  description: "Fiber internet",
  isArchived: false,
  createdAt: "2026-06-10T00:00:00Z",
  updatedAt: "2026-06-10T00:00:00Z",
};

export const dashboardFixture: DashboardSummary = {
  totalBalanceMinor: 125_000,
  monthlyIncomeMinor: 200_000,
  monthlyExpenseMinor: 50_000,
  monthlyNetMinor: 150_000,
  accounts: [{ id: accountFixture.id, name: accountFixture.name, balanceMinor: 125_000 }],
  recentTransactions: [
    {
      amountMinor: expenseTransactionFixture.amountMinor,
      transactionType: expenseTransactionFixture.transactionType,
      accountName: accountFixture.name,
      categoryName: expenseCategoryFixture.name,
      description: expenseTransactionFixture.description,
      transactionDate: expenseTransactionFixture.transactionDate,
    },
  ],
  activeBudgets: [
    {
      name: budgetFixture.name,
      categoryName: budgetFixture.categoryName,
      amountMinor: budgetFixture.amountMinor,
      spentMinor: budgetFixture.spentMinor,
      remainingMinor: budgetFixture.remainingMinor,
      progressPercentage: budgetFixture.progressPercentage,
      isNearLimit: budgetFixture.isNearLimit,
      isExceeded: budgetFixture.isExceeded,
    },
  ],
  upcomingRecurringBills: [
    {
      name: recurringBillFixture.name,
      amountMinor: recurringBillFixture.amountMinor,
      nextDueDate: recurringBillFixture.nextDueDate,
      daysRemaining: 5,
    },
  ],
  activeSavingsGoals: [
    {
      name: savingsGoalFixture.name,
      targetAmountMinor: savingsGoalFixture.targetAmountMinor,
      currentAmountMinor: savingsGoalFixture.currentAmountMinor,
      remainingAmountMinor: savingsGoalFixture.remainingAmountMinor,
      progressPercent: savingsGoalFixture.progressPercent,
    },
  ],
};

export function emptyDashboardFixture(): DashboardSummary {
  return {
    totalBalanceMinor: 0,
    monthlyIncomeMinor: 0,
    monthlyExpenseMinor: 0,
    monthlyNetMinor: 0,
    accounts: [],
    recentTransactions: [],
    activeBudgets: [],
    upcomingRecurringBills: [],
    activeSavingsGoals: [],
  };
}

File: src-tauri/tests/common/fixtures.rs
use wallet_lib::domain::accounts::dto::CreateAccountRequest;
use wallet_lib::domain::accounts::model::Account;
use wallet_lib::domain::accounts::service::AccountService;
use wallet_lib::domain::budgets::dto::CreateBudgetRequest;
use wallet_lib::domain::budgets::model::Budget;
use wallet_lib::domain::budgets::service::BudgetService;
use wallet_lib::domain::categories::dto::CreateCategoryRequest;
use wallet_lib::domain::categories::model::Category;
use wallet_lib::domain::categories::service::CategoryService;
use wallet_lib::domain::recurring_bills::dto::CreateRecurringBillRequest;
use wallet_lib::domain::recurring_bills::model::RecurringBill;
use wallet_lib::domain::recurring_bills::service::RecurringBillService;
use wallet_lib::domain::savings_goals::dto::CreateSavingsGoalRequest;
use wallet_lib::domain::savings_goals::model::SavingsGoal;
use wallet_lib::domain::savings_goals::service::SavingsGoalService;
use wallet_lib::domain::transactions::dto::CreateTransactionRequest;
use wallet_lib::domain::transactions::model::Transaction;
use wallet_lib::domain::transactions::service::TransactionService;
use wallet_lib::errors::app_error::AppError;

pub fn account_request(name: &str) -> CreateAccountRequest {
    CreateAccountRequest {
        name: name.to_string(),
        account_type: Some("cash".to_string()),
        currency: Some("MAD".to_string()),
        initial_balance_minor: Some(10_000),
    }
}

pub fn category_request(name: &str, category_type: &str) -> CreateCategoryRequest {
    CreateCategoryRequest {
        name: name.to_string(),
        category_type: category_type.to_string(),
        icon: None,
        color: None,
    }
}

pub fn transaction_request(
    account_id: &str,
    category_id: &str,
    transaction_type: &str,
    amount_minor: i64,
) -> CreateTransactionRequest {
    CreateTransactionRequest {
        account_id: account_id.to_string(),
        category_id: category_id.to_string(),
        transaction_type: transaction_type.to_string(),
        amount_minor,
        description: Some("fixture transaction".to_string()),
        transaction_date: "2026-06-10".to_string(),
    }
}

pub fn budget_request(category_id: &str, month: i64, year: i64) -> CreateBudgetRequest {
    CreateBudgetRequest {
        name: "Groceries budget".to_string(),
        category_id: category_id.to_string(),
        amount_minor: 50_000,
        month,
        year,
    }
}

pub fn recurring_bill_request(account_id: &str, category_id: &str) -> CreateRecurringBillRequest {
    CreateRecurringBillRequest {
        name: "Internet".to_string(),
        account_id: account_id.to_string(),
        category_id: category_id.to_string(),
        amount_minor: 2_500,
        frequency: "monthly".to_string(),
        next_due_date: "2026-06-15".to_string(),
        description: None,
    }
}

pub fn savings_goal_request(name: &str) -> CreateSavingsGoalRequest {
    CreateSavingsGoalRequest {
        name: name.to_string(),
        target_amount_minor: 100_000,
        current_amount_minor: Some(10_000),
        deadline_date: Some("2026-12-31".to_string()),
    }
}

pub async fn create_account(pool: &sqlx::SqlitePool) -> Result<Account, AppError> {
    AccountService::create(pool, account_request("Checking")).await
}

pub async fn create_income_category(pool: &sqlx::SqlitePool) -> Result<Category, AppError> {
    CategoryService::create(pool, category_request("Salary", "income")).await
}

pub async fn create_expense_category(pool: &sqlx::SqlitePool) -> Result<Category, AppError> {
    CategoryService::create(pool, category_request("Groceries", "expense")).await
}

pub async fn create_income_transaction(
    pool: &sqlx::SqlitePool,
    account_id: &str,
    category_id: &str,
) -> Result<Transaction, AppError> {
    TransactionService::create(
        pool,
        transaction_request(account_id, category_id, "income", 20_000),
    )
    .await
}

pub async fn create_expense_transaction(
    pool: &sqlx::SqlitePool,
    account_id: &str,
    category_id: &str,
) -> Result<Transaction, AppError> {
    TransactionService::create(
        pool,
        transaction_request(account_id, category_id, "expense", 3_500),
    )
    .await
}

pub async fn create_budget(pool: &sqlx::SqlitePool, category_id: &str) -> Result<Budget, AppError> {
    BudgetService::create(pool, budget_request(category_id, 6, 2026)).await
}

pub async fn create_recurring_bill(
    pool: &sqlx::SqlitePool,
    account_id: &str,
    category_id: &str,
) -> Result<RecurringBill, AppError> {
    RecurringBillService::create(pool, recurring_bill_request(account_id, category_id)).await
}

pub async fn create_savings_goal(pool: &sqlx::SqlitePool) -> Result<SavingsGoal, AppError> {
    SavingsGoalService::create(pool, savings_goal_request("Emergency fund")).await
}
